File: 백준/Gold/10473. 인간 대포/인간 대포.java
import java.util.*;
import java.io.*;

public class Main {

    static double sx,sy,lx,ly;
    static int n;

    static double DIS = 50.000000;

    static Node[] edges;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine()," ");
        sx = Double.parseDouble(st.nextToken());
        sy = Double.parseDouble(st.nextToken());

        st = new StringTokenizer(br.readLine()," ");
        lx = Double.parseDouble(st.nextToken());
        ly = Double.parseDouble(st.nextToken());

        n = Integer.parseInt(br.readLine());

        edges = new Node[n+2];

        edges[0] = new Node(sx,sy);

        for(int i = 1 ; i <= n ; i++) {
            st = new StringTokenizer(br.readLine()," ");
            double x = Double.parseDouble(st.nextToken());
            double y = Double.parseDouble(st.nextToken());

            edges[i] = new Node(x,y);
        }

        edges[n+1] = new Node(lx,ly);

        System.out.println(go());

        br.close();
    }

    static double go() {
        double[] dis = new double[n+2];
        for(int i = 1; i <= n+1 ; i++) {
            dis[i] = Double.MAX_VALUE;
        }

        PriorityQueue<Point> pq = new PriorityQueue<>(
                (x,y) -> Double.compare(x.sec,y.sec)
        );

        for(int i = 1 ; i <= n + 1 ; i++) {
            double tmpDis = getDis(sx,sy,edges[i].x,edges[i].y);

            dis[i] = Math.min(dis[i],tmpDis / 5.0);

            if(i <= n) {
                pq.add(new Point(i,dis[i]));
            }
        }

        while(!pq.isEmpty()) {
            Point cur = pq.poll();

            for(int i = 1 ; i <= n + 1 ; i++) {
                if(cur.idx == i)
                    continue;

                double tmpSec = 0;
                double tmpDis = getDis(edges[cur.idx].x,edges[cur.idx].y,edges[i].x,edges[i].y);
                double wSec = tmpDis / 5.0;

                if(tmpDis > DIS) {
                    tmpSec += 2.0;
                    tmpDis -= DIS;
                    tmpSec += (tmpDis / 5.0);
                }else if(tmpDis == DIS) {
                    tmpSec += 2.0;
                }else{
                    tmpSec += 2.0;
                    double tmpDis2 = DIS - tmpDis;
                    tmpSec += (tmpDis2 / 5.0);
                }

                if(dis[i] > cur.sec + Math.min(wSec,tmpSec)) {
                    dis[i] = cur.sec + Math.min(wSec,tmpSec);

                    if(i <= n) {
                        pq.add(new Point(i,dis[i]));
                    }
                }
            }
        }

        return dis[n+1];
    }

    static void print(double[] dis) {
        for(int i = 1 ; i <= n + 1; i++) {
            System.out.print(dis[i]+" ");
        }
        System.out.println();
    }


    static double getDis(double x,double y,double x1,double y1) {
        double tmpX = Math.pow(x - x1,2);
        double tmpY = Math.pow(y - y1,2);

        return Math.sqrt(tmpX + tmpY);
    }

    static class Point {
        int idx;
        double sec;

        Point(int idx,double sec) {
            this.idx = idx;
            this.sec = sec;
        }
    }

    static class Node {
        double x;
        double y;

        Node(double x,double y) {
            this.x = x;
            this.y = y;
        }
    }
}

File: 백준/Gold/10473. 인간 대포/인간 대포.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.math.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))

    var st = StringTokenizer(br.readLine(), " ")
    val sx = st.nextToken().toDouble()
    val sy = st.nextToken().toDouble()

    st = StringTokenizer(br.readLine(), " ")
    val lx = st.nextToken().toDouble()
    val ly = st.nextToken().toDouble()

    val n = br.readLine().toInt()

    val edges = Array(n + 2) { MyNode(0.0, 0.0) }
    edges[0] = MyNode(sx, sy)

    for (i in 1..n) {
        st = StringTokenizer(br.readLine(), " ")
        val x = st.nextToken().toDouble()
        val y = st.nextToken().toDouble()
        edges[i] = MyNode(x, y)
    }

    edges[n + 1] = MyNode(lx, ly)

    println(go(edges, n, sx, sy))

    br.close()
}

fun go(edges: Array<MyNode>, n: Int, sx: Double, sy: Double): Double {
    val dis = DoubleArray(n + 2) { Double.MAX_VALUE }
    dis[0] = 0.0 // Start node's distance to itself is 0

    val pq = PriorityQueue<MyPoint>(compareBy { it.sec })

    for (i in 1..n + 1) {
        val tmpDis = getDis(sx, sy, edges[i].x, edges[i].y)
        dis[i] = min(dis[i], tmpDis / 5.0)
        if (i <= n) {
            pq.add(MyPoint(i, dis[i]))
        }
    }

    while (pq.isNotEmpty()) {
        val cur = pq.poll()

        for (i in 1..n + 1) {
            if (cur.idx == i) continue

            val tmpDis = getDis(edges[cur.idx].x, edges[cur.idx].y, edges[i].x, edges[i].y)
            var tmpSec = 0.0
            val wSec = tmpDis / 5.0

            if (tmpDis > 50.0) {
                tmpSec += 2.0
                tmpSec += (tmpDis - 50.0) / 5.0
            } else if (tmpDis == 50.0) {
                tmpSec += 2.0
            } else {
                tmpSec += 2.0
                tmpSec += (50.0 - tmpDis) / 5.0
            }

            if (dis[i] > cur.sec + min(wSec, tmpSec)) {
                dis[i] = cur.sec + min(wSec, tmpSec)
                if (i <= n) {
                    pq.add(MyPoint(i, dis[i]))
                }
            }
        }
    }

    return dis[n + 1]
}

fun getDis(x: Double, y: Double, x1: Double, y1: Double): Double {
    val tmpX = (x - x1).pow(2)
    val tmpY = (y - y1).pow(2)
    return sqrt(tmpX + tmpY)
}

data class MyPoint(val idx: Int, val sec: Double)
data class MyNode(val x: Double, val y: Double)
